Try: Make site editor responsive.

jasmussen opened the pull request:
The site editor is currently not usable on mobile:

<img width="571" alt="before" src="https://user-images.githubusercontent.com/1204802/95733300-3afef000-0c82-11eb-837b-3c30689c74c5.png">

This PR does a number of things to improve that:

- It fixes the white screen issue
- It hides pinned items entirely on mobile — these modal toggles are available in the More Options menu.
- It collapses the label, so it says "Update" instead of "Update Design", on mobile.
- It hides the same undo/redo/tools/block selector buttons on the site editor, that are hidden on the post editor. 

This improves things a fair bit:

<img width="630" alt="Screenshot 2020-10-12 at 11 56 57" src="https://user-images.githubusercontent.com/1204802/95733504-887b5d00-0c82-11eb-80ed-70c6ebc36d46.png">

I wouldn't call this the end of the road, there's a great deal of polish that reains possible:

- The centered controls which are still in a design flux, probably shouldn't be optimized until we know how they will look like.
- There's some component usage that is treated differently based on post vs. site editor, this could be reduced and simplified.

But for now, this is a decent start.

jasmussen committed:
Make the site editor visible on mobile.

jasmussen committed:
Make the site editor visible on mobile.

jasmussen committed:
Hide pinned until after medium breakpoint.

jasmussen committed:
Collapse Update Design label on mobile.

jasmussen committed:
Limit tools on mobile.

jasmussen commented:
This addresses feedback from https://github.com/WordPress/gutenberg/pull/25134#issuecomment-690629371.

MichaelArestad commented on packages/interface/src/components/interface-skeleton/style.scss:
```diff
@@ -86,11 +85,13 @@ html.interface-interface-skeleton__html-container {
 	left: 0;
 	background: $white;
 	color: $gray-900;
+	width: 0;
```
I think this is really close. Something about this style is breaking the inspector panel on both the site editor and the post editor. You can see that here:

<img width="405" alt="image" src="https://user-images.githubusercontent.com/1123119/95783310-d1d1a980-0c8e-11eb-9c17-0a21c71cdc29.png">

jasmussen replied:
Awesome catch, I'll need to address that! Thank you.

jasmussen replied:
Alright this should be fixed.

jasmussen committed:
Fix regressions.

jasmussen committed:
Fix issue with sidebar not working.

jasmussen committed:
Fix issue with sidebar not working.

jasmussen commented:
Alright this one should be ready to go.

It would be really nice to get this one in soon, both so the experience isn't totally broken on mobile, but also so that actually optimizing for and considering mobile can be easier as development moves on.

david-szabo97 reviewed (commented):
The inserter panel and name of the panel are not showing up for me.

![image](https://user-images.githubusercontent.com/2256104/97169352-352d0280-178a-11eb-9987-40e65cc219cb.png)

jasmussen commented:
Argh how could I miss that !?

Taking a look, thank you!

jeyip committed:
Add desktop affordances to site editor header

Flex bases of 335px were added to the left and right site editor header toolbars to prevent their children from collapsing when shrinking the viewport. This, however, did not account for tbehavior at smaller screen widths (tablet / mobile), and caused layout issues.

To address this, we use breakpoints to define the 335px flex bases as desktop specific affordances.

jeyip commented:
I just pushed a commit to address [this comment](https://github.com/WordPress/gutenberg/pull/25884#discussion_r511776113), which seemed to be causing layout issues for mobile viewports. The header at smaller viewport widths should now look like this:

<img width="498" alt="Screen Shot 2020-10-27 at 1 58 21 AM" src="https://user-images.githubusercontent.com/5414230/97280131-1a45a700-17f9-11eb-81de-7e51703012b7.png">

I pushed it here because I thought it would be easier to test, discuss, and potentially include with the fixes made in this pull request, but I'm more than happy to rebase and remove the commit and include it in separate PR.

jasmussen committed:
Remove the width property.

jasmussen commented:
Awesome.

There are a lot of good CSS changes in this PR that I believe will hold us over for a while wrt the mobile site editor. But it's still blocked on the white screen, and I think I'll need some help getting that working. Behold:

![white](https://user-images.githubusercontent.com/1204802/97288312-f9577300-1845-11eb-997c-fbbe318bbbe0.gif)

That white screen is actually the "left sidebar" being open, all the time.

jeyip committed:
Refactor LeftSidebar rendering logic for InterfaceSkeleton

The InterfaceSkeleton component coordinates the general layout and styling of the site editor and post editor pages. It expects specific behavior from its props. For example, if the LeftSidebar is open, InterfaceSkeleton expects to be passed JSX markup as the leftSidebar prop. Otherwise, if the LeftSidebar isn't open, InterfaceSkeleton expects a null or falsey leftSidebar prop. InterfaceSkeleton styling depends on this behavior.

In edit site, the LeftSidebar prop passed into InterfaceSkeleton was always a truthy value. As a result, InterfaceSkeleton was improperly accounting for a LeftSidebar, even when it was unmounted from the screen.

These changes move the logic to determine when the sidebar should "open" up from the LeftSidebar component itself to its parent. This allows us to pass JSX markup to the InterfaceSkeleton when the inserter is open, and a falsey value when the inserter is closed.

jeyip committed:
Remove unnecessary LeftSidebar component

jeyip committed:
Refactor LeftSidebar panel styling

jeyip commented:
> There are a lot of good CSS changes in this PR that I believe will hold us over for a while wrt the mobile site editor. But it's still blocked on the white screen, and I think I'll need some help getting that working. Behold:

I pushed changes and the WSOD should now be addressed. Would love to hear from @david-szabo97 @Copons @Addison-Stavlo @vindl if there are any concerns. I moved the logic to handle conditional rendering of the Left Sidebar up to its parent, `edit-site/editor`

![2020-10-27 20 13 28](https://user-images.githubusercontent.com/5414230/97386510-372db900-1891-11eb-9686-4b29eb455226.gif)

jasmussen commented:
_Stellar_ work, @jeyip, I'm seeing this, and it's a substantial improvement:

![featured](https://user-images.githubusercontent.com/1204802/97411997-3df41480-1901-11eb-8dd3-b84135556d73.gif)

Thanks so much, I hope we can land this soon!

david-szabo97 commented:
> I pushed changes and the WSOD should now be addressed. Would love to hear from @david-szabo97 @Copons @Addison-Stavlo @vindl if there are any concerns. I moved the logic to handle conditional rendering of the Left Sidebar up to its parent, `edit-site/editor`

Any reason to move it to the parent? I prefer it to be in a separate component so less clutter in the `editor`. Opening the Inserter will now re-render the whole `Editor` if it's moved to the parent.

jeyip commented:
> Any reason to move it to the parent?

Good question!

It's because the `InterfaceSkeleton` component is designed in such a way that it expects us to pass in a false-y value when an interface element (like leftSidebar, header, footer, etc.) should not be rendered. 

https://github.com/WordPress/gutenberg/blob/7778a3630356c3a26c8e91a1bba98f7bacbb3e80/packages/interface/src/components/interface-skeleton/index.js#L87-L96

In edit-site, however, we had the left sidebar set up in such a way that it was _always_ truthy

https://github.com/WordPress/gutenberg/blob/7778a3630356c3a26c8e91a1bba98f7bacbb3e80/packages/edit-site/src/components/editor/index.js#L212-L219

This was leading to the unexpected behavior we were seeing with the mobile wsod. The `InterfaceSkeleton` was _always_ rendering a leftSidebar wrapper, even when the inserter wasn't "open". 

> Opening the Inserter will now re-render the whole Editor if it's moved to the parent.

Hmmm...this is a good point. I'll try and think about alternative approaches, but would also love to hear if you have any ideas @david-szabo97

jasmussen commented:
The post editor handles this same problem, worth noting. So we can look there for inspiration. 

Not sure if better, but it can be handled with CSS of we have a class name to indicate whether the left sidebar is open or not.

But before we go down a rabbit hole it really seems worth it to compare with the post editor.

jeyip commented:
> The post editor handles this same problem, worth noting. So we can look there for inspiration.

Great idea! I ended up comparing our implementation with the post editor's implementation, and it looks like the post editor itself would run into the same issue of unnecessary re-renders, although I might be missing something.

> Not sure if better, but it can be handled with CSS of we have a class name to indicate whether the left sidebar is open or not.

Also gave this some thought earlier, but I think we'd have to pass more information into the InterfaceSkeleton which would also cause unnecessary re-renders. 🤔

jasmussen commented:
I wonder if such a skeleton refactor should block this PR from landing, given:

- the site editor is currently unusable on < 600px
- your approach is "no worse" than what's been shipping for many months in the post editor

In that vein, we could consider this a regression fix with an opportunity to refactor the skeletons for the benefit of both editors?

jeyip commented:
Ah I should have clarified @jasmussen. I completely agree with your sentiments. I think, in the worst case scenario, we should move forward with the changes in this PR for the reasons you listed above. 

I was interested in hearing from David to see if he had ideas about solutions that would meet both criteria of:
- Avoiding a rerender of the entire editor
- Addressing the mobile wsod

david-szabo97 commented:
Since the implementation is identical to the Post Editor, I think we are good to go. I can't think of any easy way to fix the rerender for now.

We will definitely need to rethink the `InterfaceSkeleton` and `Editor`. I don't think they are going to be bottlenecks for a while but they are the root of the whole Gutenberg so it'd be great to fix them up a little bit to avoid all that unnecessary rerenders.

david-szabo97 reviewed (approved):
One thing I noticed, we can open the Navigation sidebar by clicking on the name of the current template then clicking on "Browse templates". This opens the sidebar but there is no way to close it.

The name of the template in the header also gets pushed and it is no longer visible. (If you resize while the sidebar is open you will see that the name of the template is showing up once your browser is larger.)

![image](https://user-images.githubusercontent.com/2256104/98213175-f72b9d80-1f44-11eb-839d-0093856a1840.png)

**After all,** since this PR has been going on for a while, we would have a good foundation once this PR is merged, then we can address the issues coming up.

jasmussen commented:
Thank you for the review David! I would agree there's very ripe ground for refactoring this, and it should be done. 

> One thing I noticed, we can open the Navigation sidebar by clicking on the name of the current template then clicking on "Browse templates". This opens the sidebar but there is no way to close it.

This observation, I think, is one of the best reasons for moving ahead with this, even if it isn't optimal. This is the sort of bugfix that has to help inform the sidebar work, which we can only discover when it's in some state of functional.

I'm going to give this a rebase, and see if that helps the checks pass.

jeyip commented:
I rebased master again

jeyip commented:
All specs are now passing 👍 Going to go ahead and merge the PR.
